src/thingset.h

```diff
@@ -323,17 +338,390 @@ typedef struct TsDataNode {
+ * @param pub_ch Flag to select publication channel (must match pubsub of data node)
+ * @param can_dev_id Device ID on the CAN bus
+ * @param msg_id reference to can message id storage
+ * @param msg_data reference to the buffer where the publication message should be stored
+ *
+ * @returns Actual length of the message_data or -1 if not encodable / in case of error
+ */
+int ts_bin_pub_can(ts_object_t *ts, int *start_pos, uint16_t pub_ch, uint8_t can_dev_id,
+                   uint32_t *msg_id, uint8_t *msg_data);
 
-#if CONFIG_THINGSET_CPP
+/**
+ * Update data nodes based on values provided in payload data (e.g. from other pub msg)
+ *
+ * @param ts Pointer to ts_object_t type ThingSet object.
+ * @param cbor_data Buffer containing key/value map that should be written to the data nodes
+ * @param len Length of the data in the buffer
+ * @param auth_flags Authentication flags to be used in this function (to override _auth_flags)
+ * @param sub_ch Subscribe channel (as bitfield)
+ *
+ * @returns ThingSet status code
+ */
+int ts_bin_sub(ts_object_t *ts, uint8_t *cbor_data, size_t len, uint16_t auth_flags,
+               uint16_t sub_ch);
 
-#include "thingset_cpp.h"
+/**
+ * Get data node by ID
+ *
+ * @param ts Pointer to ts_object_t type ThingSet object.
+ * @param id Node ID
+ *
+ * @returns Pointer to data node or NULL if node is not found
+ */
+TsDataNode *const ts_get_node_by_id(ts_object_t *ts, ts_node_id_t id);
+
+/**
+ * Get data node by name
+ *
+ * As the names are not necessarily unique in the entire data tree, the parent is needed
+ *
+ * @param ts Pointer to ts_object_t type ThingSet object.
+ * @param name Node name
+ * @param len Length of the node name
+ * @param parent Node ID of the parent or -1 for global search
+ *
+ * @returns Pointer to data node or NULL if node is not found
+ */
+TsDataNode *const ts_get_node_by_name(ts_object_t *ts, const char *name, size_t len, int32_t parent);
+
+/**
+ * Get the endpoint node of a provided path
+ *
+ * @param ts Pointer to ts_object_t type ThingSet object.
+ * @param path Path with multiple node names separated by forward slash
+ * @param len Length of the entire path
+ *
+ * @returns Pointer to data node or NULL if node is not found
+ */
+TsDataNode *const ts_get_endpoint(ts_object_t *ts, const char *path, size_t len);
+
+#ifdef __cplusplus
 
 #if CONFIG_THINGSET_LEGACY
 /* compatibility to legacy CPP interface */
 typedef TsArrayInfo ArrayInfo;
 typedef TsDataNode DataNode;
 #endif
 
-#endif
+} /* extern 'C' */
+
+/**
+ * Main ThingSet class
+ *
+ * Stores and handles all data exposed to different communication interfaces
+ */
+class ThingSet
+{
+public:
+    /**
+     * Initialize a ThingSet object
+     *
+     * @param data Pointer to array of TsDataNode type containing the entire node database
+     * @param num Number of elements in that array
+     */
+    inline ThingSet(TsDataNode *data, size_t num)
+    {
+        (void)ts_init(&ts, data, num);
+    };
```

Comment:
In order to avoid duplication of the comments (and the risk to forget updating them in both locations after changes) I suggest to keep the C++ shim layer simple without any comments and just add one comment at the top of the class definitions to point the user to the C functions for further information.